### PR TITLE
Roadmap.md: Discuss the near term future of rust-libp2p

### PR DESCRIPTION
**Pull request by mxinden**
Here is my (@mxinden) proposal for a roadmap for the next 6 months for the rust-libp2p maintainers (@elenaf9, @thomaseizinger, @jxs and @mxinden) working on the rust-libp2p project.

(rust-)libp2p is an open source and open community project. Thus:

- I want to make sure we listen to the community and align our work accordingly. I invite **everyone** to leave feedback here.
- The below is only planning work for the maintainers, not the community. Folks from the community may suggest and pick up work not prioritized (e.g. [Kademlia efficient querying](#kademlia-efficient-querying)) nor mentioned on this roadmap and we (maintainers) should be supporting them as best as we can.

I consider this more of a short-lived document, enabling everyone working and using rust-libp2p to discuss and eventually align on a common mission, making prioritization explicit. Once done, it will likely be outdated quickly and thus I suggest never actualy merge it. In my eyes, continued project management is better done via GitHub issues and thus all the information here should eventually make it into the corresponding GitHub issues if not already the case.

Also see:
- [libp2p roadmap](https://github.com/libp2p/specs/blob/master/ROADMAP.md)
- [go-libp2p roadmap](https://github.com/libp2p/go-libp2p/pull/1784)

The items in this roadmap are ordered by importance where I define importance as:

- Enable others to build with rust-libp2p before building components ourself.
  - e.g. cross behaviour communication over Kademlia client mode where the former enables the latter
- Invest into tools that enable us to work more efficiently
  - e.g. Testground and release process]
- Better do one thing well than many things close to right.
  - e.g. deprioritize BitSwap implementation and continue on[QUIC even once first ieration is merged. That does not imply that we won't support community members implementing it.
- Improve existing components before introducing new ones.
  - e.g. Kademlia client mode before WebTransport
 
I am curious what folks think and again encourage everyone to share their thoughts.

### CODE REVIEW

File: ROADMAP.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    - [Muxer handshake optimization](#muxer-handshake-optimization)
+    - [Bitswap implementation](#bitswap-implementation)
+    - [WebTransport](#webtransport)
+
+<!-- markdown-toc end -->
+
+## Testground abstraction
+
+Status: todo
+
```
Comment (p-shahi) [inline ROADMAP.md]:
```suggestion
Target completion: <End of Q4/Early Q1/Mid Q1/End of Q1/etc.>

```
I strongly suggest adding a target completion in each item. The roadmaps intention is to communicate the milestones within each project i.e. when we expect to ship improvements or features. A target completion will help set up expectations with external teams/collaborators/users.

We can take a similar approach to go-libp2p where each roadmap item was deconstructed into smaller deliverables and with an estimated target completion:
https://www.notion.so/Milestone-Roadmap-a90d888d6bf14b999131ea7a802c9960 (It's in Notion now but will make its way back to markdown document today)

File: ROADMAP.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    - [swarm/handler: Decouple ConnectionHandler from {In,Out}boundUpgrade #2863](#swarmhandler-decouple-connectionhandler-from-inoutboundupgrade-2863)
+    - [Generic connection management](#generic-connection-management)
+    - [Kademlia efficient querying](#kademlia-efficient-querying)
+    - [Kademlia client mode](#kademlia-client-mode)
+    - [Hole punching](#hole-punching)
+    - [Streaming response protocol aka. the ultimate escape hatch](#streaming-response-protocol-aka-the-ultimate-escape-hatch)
+    - [Improved WASM support](#improved-wasm-support)
+    - [Muxer handshake optimization](#muxer-handshake-optimization)
+    - [Bitswap implementation](#bitswap-implementation)
+    - [WebTransport](#webtransport)
```
Comment (p-shahi) [inline ROADMAP.md:24]: There are 17 items here, do you think it's feasible to ship all of them in the next two quarters? For ones that are lower priority and will likely not get shipped in the next two quarters, I suggest omitting them from this document

File: ROADMAP.md (hunk 1 above)
Comment (p-shahi) [inline ROADMAP.md]:
```suggestion
Priority: <P0/P1/P2>

```
Suggest adding a relative priority to each roadmap item here as well. It is hard to discern which items are most critical vs which are nice to haves.

File: ROADMAP.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+We are currently implementing WebRTC for **browser-to-server** connectivity in
+https://github.com/libp2p/rust-libp2p/pull/2622. More specifically the server side. This will enable
+browser nodes to connect to rust-libp2p nodes where the latter only have self-signed TLS
+certificates. See https://github.com/libp2p/specs/pull/412 for in-depth motivation.
+
+Long term we should enable rust-libp2p running in the browser via WASM to use the browser's WebRTC
+stack. Though that should only happen after improved WASM support, see below.
+
+## Release process
```
Comment (p-shahi) [inline ROADMAP.md]: I would omit this from the roadmap as this item can be characterized as "ongoing project maintenance work."

**Review by p-shahi — COMMENTED** [review]
This is awesome, I just gave some drive by comments.

File: ROADMAP.md (hunk 2 above)
Comment (thomaseizinger) [inline ROADMAP.md]: All of these are exciting!

File: ROADMAP.md [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+Tracking: https://github.com/libp2p/rust-libp2p/issues/2632
+
+I think this is a common component that many users need to build peer-to-peer applications. In
+addition, it is very performance critical and thus likely challenges many of our existing designs in
+rust-libp2p.
+
+I would prioritize it below [Muxer handshake optimization](#muxer-handshake-optimization) following
+the convention of improving existing components over introducing new ones. Users have and can
+implement their own implementations and are thus not blocked on the rust-libp2p project.
```
Comment (thomaseizinger) [inline ROADMAP.md]: Why included these here if they are low priority?

**Review by thomaseizinger — APPROVED** [review]
> I consider this more of a short-lived document, enabling everyone working and using rust-libp2p to discuss and eventually align on a common mission, making prioritization explicit. Once done, it will likely be outdated quickly and thus I suggest never actualy merge it.

Happy with that!

File: ROADMAP.md (hunk 1 above)
Comment (BigLep) [reply ROADMAP.md]: Agreed.  Given an indication of what we're thinking will land when is important for stakeholders.  Monthly granularity is good.  If you need to do quarters, that's fine.  Without some estimated dates, it's hard for a stakeholder to make plans.  For example, if I see something I really care about that is in the distant future, I at least know that and can respond accordingly (e.g., wait, jump in to bring it in, see if can convince about prioritization).

A brief mapping of milestones to dates like to go-libp2p does here works: https://github.com/libp2p/go-libp2p/pull/1804

File: ROADMAP.md (hunk 2 above)
Comment (BigLep) [reply ROADMAP.md:24]: While not critical, I think it's ok/good to have them in the document, but it should be clear that they are unplanned for the next 6 month.  The value in keeping them in my opinion is:
1. it gives clarity to what you said "no" to.  If someone cares about one of those initiatives, it's clear that we thought about it and chose no.
2. It gives a nod to things we'd consider for the future.

Feel free to dump them as well though if that's easier.

File: ROADMAP.md (hunk 3 above)
Comment (BigLep) [reply ROADMAP.md]: Agreed that this isn't something that belongs on the stakeholder roadmap.  Per https://docs.google.com/document/d/1XJiFG9tTCDi4RsKEAivHgcXueIqqwhkJDE-OKvy8yec/edit#heading=h.vwnoff9vzgnf, "Each milestone should be scoped to a real-world, externalizable deliverable - something that matters to your users or collaborators in a notable way"

That said, getting a release process documented/established is one of the first things I would want us to do from an engineering regard.  Basically, I believe it belong at/near the top of the engineering team roadmap, but not hear for external stakeholders.

**Review by BigLep — COMMENTED** [review]
Thanks for putting this together @mxinden .  I know @p-shahi will engage more here, but wanted to give a couple of comments.

> I consider this more of a short-lived document, enabling everyone working and using rust-libp2p to discuss and eventually align on a common mission, making prioritization explicit. Once done, it will likely be outdated quickly and thus I suggest never actualy merge it. 

It's ok if this document specifically is short-lived, but I think we need a document/table somewhere that communicates to stakeholders the major initiatives of the project and when we estimate completing them as described in https://docs.google.com/document/d/1XJiFG9tTCDi4RsKEAivHgcXueIqqwhkJDE-OKvy8yec/edit# .  This "stakeholder roamdap" will be higher-level than the week-to-week plans that the engineering team is using to drive their execution.  We see this with projects like FVM where they have https://fvm.filecoin.io/#roadmap-4 but then also fine grained detailed planning to hit those milestones.

The hat I would wear is: imagine you're someone like Juan or Molly who has been told they have to share tomorrow about what rust-libp2p is up to and what users can expect in the next 6 months or if they are trying to summarize the efforts across multiple EngRes teams.  You'd want the higher-level items that really matter to collaborators and users.  

If this document is the best we have currently (which is totally fine), I do think we should get it checked in but:
1. We can communicate the document's status (actively being reviewed/updated in October)
2. We make it clear how someone leaves comments or tracks the status of additional updates.  I presume we have a tracking "2023 roadmap issue" that we link to from this document.

> In my eyes, continued project management is better done via GitHub issues and thus all the information here should eventually make it into the corresponding GitHub issues if not already the case.

I agree, but have a few comments:
1. Agreed GitHub issues are far superior for linking between other issues/prs, checklists, labeling, commenting etc.  I think each milestone item below having a tracking makes sens.e. I see you have that in some cases.  (That said, I would make sure to link to rust-libp2p items which can then link out to spec issues).
2. We shouldn't require someone to go spelunking through github issues to get a sense of the high-level roadmap.  I do think it's possible to have a world where you have all the upcoming epics on a clean board for someone to get the stakeholder view, but that will take setup.  Until that happens, I think going with a document like you have now is probably the easiest way to go.  Given the "stakeholder" roadmap is only a handful of milestones, I don't think it's a lot of duplicated work to have that be it's own document/presentation which is separate for where all the engineering work is happening day-to-day in GitHub issues.

In summary, until we have something alternative/better that communicates the direction of the project, I do think we should be driving to get this merged in.

**Comment by MarcoPolo** [comment]
I'd love to see benchmarking show up here because:

1. The sooner we do this, the sooner we can quantify the performance and make regressions harder.
2. Perception (even if unfounded) of a slow networking library pose an existential risk to libp2p.
3. It lets us highlight QUIC improvements in a concrete manner.
4. Signal to users that we are committed to a performant library. This is the "zero cost abstraction" of networking libraries (alright maybe not zero cost, but _close enough_).

File: ROADMAP.md (hunk 1 above)
Comment (mxinden) [reply ROADMAP.md]: The roadmap items already express relative priority through the order in which the items are listed. What value would an additional _priority_ property per roadmap item bring? In other words, what questions would this property help answer?

> It is hard to discern which items are most critical vs which are nice to haves.

I consider all items on this roadmap important, i.e. worth doing eventually.

File: ROADMAP.md (hunk 2 above)
Comment (mxinden) [reply ROADMAP.md:24]: I am also in favor of tracking items beyond the next 6 months.

I am adding a _target completion_ property to each item. I will specifically mark the items that I expect to be beyond the next 6 month.

File: ROADMAP.md (hunk 3 above)
Comment (mxinden) [reply ROADMAP.md]: :+1: removed with 53a74578.

File: ROADMAP.md (hunk 4 above)
Comment (mxinden) [reply ROADMAP.md]: I consider all items on this roadmap worth doing eventually. Given that we are resource constraint we have to make trade offs somewhere. That said, I do think it is worth tracking them here for future planning.

I suggest we move this discussion to https://github.com/libp2p/rust-libp2p/pull/2997#discussion_r989168789. I am resolving here.

File: ROADMAP.md (hunk 1 above)
Comment (p-shahi) [reply ROADMAP.md]: > The roadmap items already express relative priority through the order in which the items are listed. What value would an additional priority property per roadmap item bring? In other words, what questions would this property help answer?

I see since that's the case then I agree, an additional priority property would not bring any additional value.

But it wasn't evident that it was listed in order of descending priority when I first reviewed it. The addition of `Target completion:` helps.
My suggestion is to add a blurb at the top of this doc and say as much, this will help new readers out.

File: ROADMAP.md (hunk 2 above)
Comment (p-shahi) [reply ROADMAP.md:24]: Cool that makes sense.

File: ROADMAP.md (hunk 1 above)
Comment (p-shahi) [reply ROADMAP.md]: Oh nevermind you actually specified it in the description
>The items in this roadmap are ordered by importance where I define importance as:

My bad

**Comment by mxinden** [comment]
Thanks for the input everyone.

- Make this a high level document for stakeholders.
- Track detailed work in project board https://github.com/orgs/libp2p/projects/4/views/1
- Fine with it being merged into master.
- Still need to address benchmark item raised by @MarcoPolo.

**Comment by mxinden** [comment]
@dignifiedquire, @melekes, @kpp, @AgeManning, @divagant-martian, @ackintosh, @ everyone:

Want to give this a review? Anything you think is missing? Anything you would like to see prioritized?

File: ROADMAP.md (hunk 1 above)
Comment (mxinden) [reply ROADMAP.md]: > My suggestion is to add a blurb at the top of this doc and say as much, this will help new readers out.

Good idea @p-shahi. Added with https://github.com/libp2p/rust-libp2p/pull/2997/commits/aed8b29ceff9e963e494b11d27efe61c47f53f91.

**Review by melekes — APPROVED** [review]
👍

**Review by kpp — APPROVED** [review]

File: ROADMAP.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+|--------|-------------------|---------------------------------------------------|---------------------------------------------------|-----------------------------------------------|
+| todo   | Q1/2023           | https://github.com/libp2p/rust-libp2p/issues/2680 | https://github.com/libp2p/rust-libp2p/issues/2832 | [Kademlia client mode](#kademlia-client-mode) |
+
+Today `NetworkBehaviour` implementations like Kademlia, GossipSub or Circuit Relay v2 can not
+communicate with each other, i.e. can not make use of information known by another
+`NetworkBehaviour` implementation. First step would be to exchange standard information about remote
+peers (e.g. supported protocols).
+
+Long term we might consider a generic approach for `NetworkBehaviours` to exchange data. Though that
+would deserve its own roadmap item.
```
Comment (thomaseizinger) [inline ROADMAP.md]: I think this is an important item that we should work on because it will be a nice ergonomic improvement for our users. At the same time though, it is only just that: An ergonomic improvement. A user can integrate identify and kademlia today, they just need to write a bit of code for it.

With that in mind, I think we should remove this item from this roadmap.

File: ROADMAP.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+work before that.
+
+## Kademlia client mode
+
+| Status | Target Completion | Tracking                                          | Dependencies                                                    | Dependents |
+|--------|-------------------|---------------------------------------------------|-----------------------------------------------------------------|------------|
+| todo   | Q1/2023           | https://github.com/libp2p/rust-libp2p/issues/2032 | [Cross behaviour communication](#cross-behaviour-communication) |            |
+
+Kademlia client mode will enhance routing table health and thus have a positive impact on all
+Kademlia operations.
```
Comment (thomaseizinger) [inline ROADMAP.md]: I think these could go under a heading "Improving runtime efficiency".

File: ROADMAP.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+| Status | Target Completion | Tracking                                          | Dependencies | Dependents |
+|--------|-------------------|---------------------------------------------------|--------------|------------|
+| todo   | Q1/2023           | https://github.com/libp2p/rust-libp2p/issues/2824 |              |            |
+
+See https://github.com/libp2p/rust-libp2p/issues/2824 for motivation. Given that this will enable
+downstream users to easier integrate with rust-libp2p, I think this counts as a "improving existing
+components" over "introducing a new component".
+
+First draft is in https://github.com/libp2p/rust-libp2p/pull/2828
```
Comment (thomaseizinger) [inline ROADMAP.md]: My thoughts on this one are similar to cross-behaviour communication but it is a little less clear because you need to work around the fact that connections get established first and a behaviour would have to instantly kill it again. This might get the job done but is a bit hacky.

I am inclined to say that this should either also be removed or grouped under a "Developer ergonomics" heading.

File: ROADMAP.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+## Streaming response protocol aka. the ultimate escape hatch
+
+| Status | Target Completion | Tracking                                          | Dependencies | Dependents |
+|--------|-------------------|---------------------------------------------------|--------------|------------|
+| todo   | Q1/2023           | https://github.com/libp2p/rust-libp2p/issues/2657 |              |            |
+
+rust-libp2p is very opinionated on how to write peer-to-peer protocols. There are many good reasons
+for this, and I think we should not change directions here. That said, the ultimate escape hatch -
+allowing users to create a stream and do whatever they want with it - will make it easier for
+newcomers to get started.
```
Comment (thomaseizinger) [inline ROADMAP.md]: My way of looking at this is that we should make it easier for users to write `NetworkBehaviour`s instead of giving them an escape hatch, hence I think we should remove this.

What we could do is add an item to the "Developer Ergonomics" heading (if you end up introducing it) that briefly mentions "Implementations of custom protocols".

File: ROADMAP.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+| Status | Target Completion | Tracking                                                                                                                                                | Dependencies | Dependents |
+|--------|-------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------|--------------|------------|
+| todo   | Q2/2023           | Security protocol in multiaddr https://github.com/libp2p/specs/pull/353 and early muxer negotiation https://github.com/libp2p/rust-libp2p/issues/2994 |              |            |
+
+Short term, investing into rust-libp2p's QUIC support will likely give us a larger performance win,
+thus neither of the two optimizations is planned for 2022. While great to have, it has not been
+requested from any rust-libp2p users.
+
+Long term, given that this will give us a great performance gain, we should definitely tackle it. It
+also allows us to catch up and thus be consistent with go-libp2p.
```
Comment (thomaseizinger) [inline ROADMAP.md]: This could be grouped into "performance optimisations" as well.

**Review by thomaseizinger — COMMENTED** [review]
I think this is a good roadmap!

My suggestion would be to present it slightly differently. For some features (TLS, QUIC and WebRTC) it is pretty clear to outside users and stakeholders, what the outcome is.

For some of the other items, that intention is missing. I would suggest to introduce the following high-level items:

- Performance optimisations
- Developer ergonomics

I think those are goals that are easier to relate to. It is also easier to say: "In Q4, rust-libp2p will focus on TLS, QUIC, WebRTC and developer ergonomics".

On issue with this approach is that "developer ergonomics" is never complete but I don't think it needs to be. It can be on the roadmap for every quarter and just the concrete goals underneath get completed.
I think this also ties in well with @MarcoPolo's comment about benchmarks. Those can go under "Performance optimisations" as an initial starting point.

Last but not least, what is missing for me here is a concrete call-out to spec-compliance. It is easy to say that this should be implicit (i.e. all implementations always being compliant). However, if we don't give it any attention, it is easy to overlook. I'd love to have an epic for spec-compliance where we can f.e. group all the testground work in but also things like https://github.com/libp2p/specs/issues/466.

File: ROADMAP.md (hunk 5 above)
Comment (mxinden) [reply ROADMAP.md]: I would like to keep the paragraph here to stress that generic cross-behaviour-communication is **not** on the roadmap. Otherwise folks might confuse the scope of  this roadmap item.

File: ROADMAP.md (hunk 6 above)
Comment (mxinden) [reply ROADMAP.md]: I am confused. Why is Kademlia a runtime or part of a runtime? Or as in "at run time"?

File: ROADMAP.md (hunk 7 above)
Comment (mxinden) [reply ROADMAP.md]: > This might get the job done but is a bit hacky.

As an aside, that is how https://github.com/sigp/lighthouse and https://github.com/paritytech/substrate do it today.

> I am inclined to say that this should either also be removed or grouped under a "Developer ergonomics" heading.

In my eyes, this goes beyond ergonomics. I.e. it enables users to implement a clean eclipse defense.

File: ROADMAP.md (hunk 8 above)
Comment (mxinden) [reply ROADMAP.md]: > My way of looking at this is that we should make it easier for users to write `NetworkBehaviour`s instead of giving them an escape hatch, hence I think we should remove this.

I am still undecided here. I think we need to do both. I fear that we will never be able to make `NetworkBehaviour` (a) easy enough for a newcomer to implement a proof-of-concept, while (b) still supporting advanced protocols like Kademlia, Gossipsub and BitSwap.

I suggest we continue experimenting with both. I don't have an opinion on which one to prioritize.

**Comment by mxinden** [comment]
Thanks for the input @thomaseizinger!

On the matter of grouping items, I decided against it, but instead introduced categories for each item.

Reason for not grouping items is that thus the prioritization across groups gets lost. E.g. while WebRTC and WebTransport are in the _Connectivity_ category, one is high priority, the other is low priority, with many items from other groups in between.

Let me know if that works for you.

> Last but not least, what is missing for me here is a concrete call-out to spec-compliance. It is easy to say that this should be implicit (i.e. all implementations always being compliant). However, if we don't give it any attention, it is easy to overlook. I'd love to have an epic for spec-compliance where we can f.e. group all the testground work in but also things like [libp2p/specs#466](https://github.com/libp2p/specs/issues/466).

I don't have a strong opinion. We removed the Testground item from this roadmap as it is not user facing. In the same spirit I suggest not tracking specification compliance here either. How about tracking it on https://github.com/orgs/libp2p/projects/4 instead?

That is not to say that specification compliance is not a high priority.

> On issue with this approach is that "developer ergonomics" is never complete but I don't think it needs to be. It can be on the roadmap for every quarter and just the concrete goals underneath get completed.
> I think this also ties in well with @MarcoPolo's comment about benchmarks. Those can go under "Performance optimisations" as an initial starting point.

As above, I suggest not tracking benchmarking on this roadmap given that it is (arguably) not user facing.

On this matter, I would prioritize it below the testing effort along the lines of: "make it work, make it correct, make it fast".

**Comment by efrisch** [comment]
@mxinden There's a few typos in this sentence. Is this saying QUIC will continue to have higher priority than the BitSwap implementation once QUIC is merged?
>   * e.g. deprioritize BitSwap implementation and continue on[QUIC even once first ieration is merged. That does not imply that we won't support community members implementing it.

**Review by divagant-martian — COMMENTED** [review]
While this looks fairly ambitious to me, it's really great! Looking forward to it

File: ROADMAP.md (hunk 5 above)
Comment (thomaseizinger) [reply ROADMAP.md]: I am talking about the entire roadmap item, not just the generic approach :)

File: ROADMAP.md (hunk 6 above)
Comment (thomaseizinger) [reply ROADMAP.md]: The latter. A better term is probably "Performance improvements".

File: ROADMAP.md (hunk 7 above)
Comment (thomaseizinger) [reply ROADMAP.md]: Perhaps it should go under a "Security" heading then?

**Comment by thomaseizinger** [comment]
> Thanks for the input @thomaseizinger!
> 
> On the matter of grouping items, I decided against it, but instead introduced categories for each item.
> 
> Reason for not grouping items is that thus the prioritization across groups gets lost. E.g. while WebRTC and WebTransport are in the _Connectivity_ category, one is high priority, the other is low priority, with many items from other groups in between.
> 
> Let me know if that works for you.

Yeah the priority is a bit tricky. I think categories work. I guess it depends on what you think is more important to present to stake-holders:

- Priority of concrete items: Use that as the sort-order and have categories as tags.
- Broad categorization of work: List the categories and stick a priority label on each item.

I don't mind either way. Without asking some of the non-technical stakeholders, we are unlikely to know what they'd prefer.

> > Last but not least, what is missing for me here is a concrete call-out to spec-compliance. It is easy to say that this should be implicit (i.e. all implementations always being compliant). However, if we don't give it any attention, it is easy to overlook. I'd love to have an epic for spec-compliance where we can f.e. group all the testground work in but also things like [libp2p/specs#466](https://github.com/libp2p/specs/issues/466).
> 
> I don't have a strong opinion. We removed the Testground item from this roadmap as it is not user facing. In the same spirit I suggest not tracking specification compliance here either. How about tracking it on [github.com/orgs/libp2p/projects/4](https://github.com/orgs/libp2p/projects/4) instead?

I mostly mentioned it because it may be worth giving visibility to this to other stakeholders. I am certain that we will be giving it an appropriate amount of attention :)

> > On issue with this approach is that "developer ergonomics" is never complete but I don't think it needs to be. It can be on the roadmap for every quarter and just the concrete goals underneath get completed.
> > I think this also ties in well with @MarcoPolo's comment about benchmarks. Those can go under "Performance optimisations" as an initial starting point.
> 
> As above, I suggest not tracking benchmarking on this roadmap given that it is (arguably) not user facing.
> 
> On this matter, I would prioritize it below the testing effort along the lines of: "make it work, make it correct, make it fast".

:+1:

**Review by thomaseizinger — APPROVED** [review]

File: ROADMAP.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
+# rust-libp2p Roadmap
+
+Below is a high level roadmap for the rust-libp2p project. Items are ordered by priority (high to
+low).
+
+For a more detailed view see https://github.com/orgs/libp2p/projects/4/views/1.
```
Comment (BigLep) [inline ROADMAP.md:6]: A couple of things:
1. That project board is private
2. Maybe rename it to "rust-libp2p Detailed Roadmap"
3. Add a readme https://github.com/orgs/libp2p/projects/4/settings  ?  It can have a note pointing back to this document.

File: ROADMAP.md [hunk 11]
```diff
@@ -0,0 +1,189 @@
+# rust-libp2p Roadmap
+
+Below is a high level roadmap for the rust-libp2p project. Items are ordered by priority (high to
```
Comment (BigLep) [inline ROADMAP.md:3]: Additional things to potentially cover:
1. Link to the specs vision roadmap (which should ideally cover core tenets of the project like security, stability, performance, and specs first).  I suggested to @p-shahi in https://github.com/libp2p/go-libp2p/pull/1831#discussion_r998906511 to factor some of the tenets there so we can link to them from the various libp2p implementation roadmaps.
2. How does someone leave feedback or comments.  go-libp2p and testground went with a pinned tracking issue with a comment at the top: https://github.com/libp2p/go-libp2p/blob/master/ROADMAP.md#go-libp2p-roadmap-q422q123
3. I think we need to add notes about how interoperability and performance testing are critical initiatives for the project and in many cases take higher priority than other initiatives listed here.  go-libp2p has some notes on this here: https://github.com/libp2p/go-libp2p/pull/1831

File: ROADMAP.md [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    - [Bitswap implementation](#bitswap-implementation)
+    - [WebTransport](#webtransport)
+
+<!-- markdown-toc end -->
+
+## TLS support
+
+| Category     | Status      | Target Completion | Tracking                                        | Dependencies                       | Dependents |
+|--------------|-------------|-------------------|-------------------------------------------------|------------------------------------|------------|
+| Connectivity | In progress | Q4/2022           | https://github.com/libp2p/rust-libp2p/pull/2945 | [QUIC](#experimental-quic-support) |            |
```
Comment (BigLep) [inline ROADMAP.md:32]: Minor but I think "QUIC" is a dependent rather than a dependency.

File: ROADMAP.md [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+
+This allows us to secure both TCP and QUIC connections using TLS. This is a requirement for QUIC
+support. Running TLS on top of TCP is a nice to have, since we already have noise.
+
+## Experimental QUIC support
```
Comment (BigLep) [inline ROADMAP.md:38]: What will a stakehodler get frome experimental QUIC support vs. full support?  Will we not have "Transport::dial_as_listener to support hole-punching"?  Do we feel good about anyone putting production workloads on this?  

As a stakeholder, I would imagine production support and completion is what someone ultimately cares about. If we are breaking that into a separate roadmap item, is there a reason it's not on the list?   I would only imagine leaving something in an experimental state if we're not sure about it's utility or the API to expose, but we already know we need QUIC in production, so why not accelerate completing this?

File: ROADMAP.md [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+stack. Though that should only happen after improved WASM support, see below.
+
+## Cross Behaviour communication
+
+| Category             | Status | Target Completion | Tracking                                          | Dependencies                                      | Dependents                                    |
+|----------------------|--------|-------------------|---------------------------------------------------|---------------------------------------------------|-----------------------------------------------|
+| Developer ergonomics | todo   | Q1/2023           | https://github.com/libp2p/rust-libp2p/issues/2680 | https://github.com/libp2p/rust-libp2p/issues/2832 | [Kademlia client mode](#kademlia-client-mode) |
+
+Today `NetworkBehaviour` implementations like Kademlia, GossipSub or Circuit Relay v2 can not
+communicate with each other, i.e. can not make use of information known by another
```
Comment (BigLep) [inline ROADMAP.md:74]:
```suggestion
communicate with each other, i.e. cannot make use of information known by another
```

File: ROADMAP.md [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+`NetworkBehaviour` implementation. Users need to write the wiring code by hand to e.g. enable
```
Comment (BigLep) [inline ROADMAP.md:75]: I don't know: is this a big deal?  Is this common for rust-libp2p users to need to do?

Will rust-libp2p users look at this item and think, "awesome, I've been waiting for this.  This will make my application development so much easier"?

On the surface this reads like an internal refactor, and I am just checking to make sure this carries enough weight to be listed here.  

If we don't view this as an important stakeholder milestone, this can be an implementation endeavor for "Kademlia client mode"

File: ROADMAP.md [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+
+Below is a high level roadmap for the rust-libp2p project. Items are ordered by priority (high to
+low).
+
+For a more detailed view see https://github.com/orgs/libp2p/projects/4/views/1.
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
+**Table of Contents**
+
+- [rust-libp2p Roadmap](#rust-libp2p-roadmap)
```
Comment (BigLep) [inline ROADMAP.md:11]: (nit) This isn't needed/useful in the TOC.

File: ROADMAP.md [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+|----------|--------|-------------------|---------------------------------------------------|--------------|------------|
+|          | todo   |                   | https://github.com/libp2p/rust-libp2p/issues/2632 |              |            |
+
+I think this is a common component that many users need to build peer-to-peer applications. In
+addition, it is very performance critical and thus likely challenges many of our existing designs in
+rust-libp2p.
+
+I would prioritize it below [handshake optimization](#handshake-optimizations) following the
+convention of improving existing components over introducing new ones. Users have and can implement
+their own implementations and are thus not blocked on the rust-libp2p project.
```
Comment (BigLep) [inline ROADMAP.md:178]: Would other groups like numober0/iroh consider contributing this?

**Comment by mxinden** [comment]
> @mxinden There's a few typos in this sentence. Is this saying QUIC will continue to have higher priority than the BitSwap implementation once QUIC is merged?
> 
> > * e.g. deprioritize BitSwap implementation and continue on[QUIC even once first ieration is merged. That does not imply that we won't support community members implementing it.

Yes @efrisch. In my eyes we should prioritize fixing and improving existing functionality (QUIC once merged) before new functionality (BitSwap).

**Comment by mxinden** [comment]
Thanks for all the input here.

Looking forward to tackling these with all of you. Exciting times ahead.

I am merging here.

Feel free to post additional comments here or create pull requests against the file. This is a living document owned by all of us.

**Comment by mxinden** [comment]
Semver check failure is discussed in https://github.com/libp2p/rust-libp2p/pull/2647, Testground failure is tracked in https://github.com/libp2p/test-plans/issues/59.

File: ROADMAP.md (hunk 5 above)
Comment (mxinden) [reply ROADMAP.md]: I think we should prioritize it. In my eyes having this feature significantly reduces the amount of complexity forced onto new users.

**Comment by BigLep** [comment]
@mxinden : arg!  I see that I never submitted a review I did on Tuesday.  Arg - my bad.  I'm submitting it now.

File: ROADMAP.md (hunk 17 above)
Comment (mxinden) [reply ROADMAP.md:178]: We have been talking about this with them. Maybe.

Comment (mxinden) [reply ROADMAP.md:178]: See also conversation in https://github.com/libp2p/rust-libp2p/issues/2632.

File: ROADMAP.md (hunk 15 above)
Comment (mxinden) [reply ROADMAP.md:75]: In my eyes this is important beyond Kademlia. Given that all users currently have to do this manually, I do expect folks appreciating the change and looking forward to it.

File: ROADMAP.md (hunk 13 above)
Comment (mxinden) [reply ROADMAP.md:38]: Good point. With the above in mind, I suggest making the effort beyond the first iteration part of this roadmap item and thus prioritizing it.

**Comment by mxinden** [comment]
For the record, follow ups to @BigLep's review above is happening in https://github.com/libp2p/rust-libp2p/pull/3062.

File: ROADMAP.md (hunk 10 above)
Comment (BigLep) [reply ROADMAP.md:6]: @mxinden : I took care of 2 and 3.  I can't do 3 (I think because I don't have sufficient permissions.)

File: ROADMAP.md (hunk 11 above)
Comment (BigLep) [reply ROADMAP.md:3]: @mxinden : I'd like to get a callout about us focusing on interoperability testing right now.  go-libp2p has this here: https://github.com/libp2p/go-libp2p/blob/master/ROADMAP.md#benchmarking-and-testing

Comment (BigLep) [reply ROADMAP.md:3]: We can do the work here: https://github.com/libp2p/rust-libp2p/pull/3091
